dask_searchcv/adaptive.py

```diff
@@ -0,0 +1,385 @@
+            ``best_estimator_`` and ``best_params_``
+
+        history : list of dicts
+            The recorded history for this adaptive process. It includes the
+            evaluations on each step, and it's possible to use this to see each
+            decision on which models to keep evaluating. Each dictionary in this
+            variables has ``set(keys) == {}``
+
+        Notes
+        -----
+        This implements the `Hyperband model selection algorithm`_
+        and only requires one input, ``max_iter`` which is porprotional to
+        "computational effort" required. Notably it does not require balancing
+        a trade-off between the number of params to evaluate and how long to
+        train each config.
+
+        Hyperband has some theoritical guarantees on finding (close to) the
+        optimal config from ``params`` given this computational effort
+        ``max_iter``.
+
+        .. _Hyperband model selection algorithm: https://arxiv.org/abs/1603.06560
+        """
+        self.params = params
+        self.model = model
+        self.R = max_iter
+        self.eta = eta
+        self.best_val_score = -np.inf
+        self.n_iter = 0
+        self.run_in_parallel = run_in_parallel
+
+        if not hasattr(model, 'warm_start'):
+            warnings.warn('model has no attribute warm_start. Hyperband will assume it '
+                          'is reusing previous calls to `partial_fit` during each call '
+                          'to `partial_fit`')
+        else:
+            if not model.warm_start:
+                raise ValueError('Hyperband relies on model(..., warm_start=True).')
+
+        self.history = []
+
+        super_default = dict(scoring=None, iid=True, refit=True, cv=None,
+                             error_score='raise', return_train_score=_RETURN_TRAIN_SCORE_DEFAULT,
+                             scheduler=None, n_jobs=-1, cache_cv=True)
+        for k, v in super_default.items():
+            if k in kwargs and kwargs[k] != super_default[k]:
+                warnings.warn('Hyperband ignores the keyword argument {k}.')
+        super(Hyperband, self).__init__(model, **kwargs)
+
+    def fit(self, X, y, dry_run=False, verbose=False, **fit_kwargs):
+        """
+        This function implicitly assumes that higher scores are better.
+
+        This is a general interface; ``model`` can be any class that
+
+        * implements ``model.set_params(**kwargs) -> model``
+        * implements ``model.partial_fit(X, y, classes)``.
+          ``classes`` can be ``np.unique(y_total)``.
+        * implements ``model.score(X, y) -> number``
+        * supports ``sklearn.base.clone(model, safe=True)``. Support for this
+          function comes down to having an function ``get_params``.
+
+        """
+        import distributed
+        client = _get_client()
+
+        variables = {'val_score': -np.inf, 'model': None, 'config': None,
+                     'classes': np.unique(y).tolist(), 'eta': self.eta,
```

Comment:
regarding classes, Hyperband will work with regression as well, right? In that case, we don't want to `unique`.

Since unique is maybe expensive (even for categorical `y`), we'll want to require the user provide it. With e.g. `sklearn.linear_model.SGDClassifier`, you're required to pass `classes` to the first `partial_fit` call.

We have two options for where the user should provide the required fit kwargs

1. As a `fit_kwargs` to argument to `Hyperband` itself
2. As keyword arguments in `.fit` that are passed through

I'll do some research on which is preferable. Perhaps @ogrisel has thoughts.

Reply:
I've stripped `unique` out of the implementation. Right now, I have `Hyperband.fit(X, y, **fit_kwargs)`. This is what [sklearn GridSearchCV.fit][1] does, and passing `fit_kwargs` to `__init__` is deprecated.

[1]:http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV.fit